File: c++/subsets-ii.cc
class Solution {
private:
  void recursiveSubset(vector<int>& nums, vector<vector<int>>& combinations, vector<int>& cur, int index) {
    combinations.push_back(cur);
    for (int i = index; i < nums.size(); ++i) {
      cur.push_back(nums[i]);
      recursiveSubset(nums, combinations, cur, i + 1);
      cur.pop_back();
    }
  }
public:
  vector<vector<int>> subsetsWithDup(vector<int>& nums) {
    sort(nums.begin(), nums.end());
    vector<vector<int>> combinations;
    vector<int> cur;

    recursiveSubset(nums, combinations, cur, 0);
    sort(combinations.begin(), combinations.end());

    vector<vector<int>> uniqueCombinations;
    uniqueCombinations.push_back(combinations[0]);

    for (int i = 1; i < combinations.size(); ++i) {
      if (uniqueCombinations.back() == combinations[i]) continue;
      uniqueCombinations.push_back(combinations[i]);
    }

    return uniqueCombinations;
  }
};

File: python/subsets-ii.py
class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        subsets = [[]]
        nums.sort()

        def recursiveBacktracking(startIndex: int = 0, currSet: List[int] = []):
            for index in range(startIndex, len(nums)):
                if index > startIndex and nums[index] == nums[index - 1]:
                    continue
                currSet.append(nums[index])
                subsets.append(currSet[:])
                recursiveBacktracking(index + 1, currSet[:])
                currSet.pop()
        recursiveBacktracking()
        return subsets
